### PR TITLE
feat: add  dry-run flag to snapshot export

### PR DESCRIPTION
**Pull request by hanabi1224**
## Summary of changes

Changes introduced in this pull request:
- for detailed task description see https://github.com/ChainSafe/forest/issues/2550
- the same flag has been implemented and tested as part of https://github.com/ChainSafe/forest/pull/2486  

## Reference issue to close (if applicable)

Closes https://github.com/ChainSafe/forest/issues/2550

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG](https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md) is up-to-date. All user-facing changes should be reflected in this document.

### CODE REVIEW

**Review by LesnyRumcajs — COMMENTED** [review]
I don't particularly like the `if dry_run else` approach. It makes the code difficult to read and maintain, with too many branches to follow. Cyclomatic complexity & stuff.

Have you considered adding a dummy writer to the `export` method? It would make the logic more straightforward. I don't know if such implementation exists somewhere as crates, but it should be easy to create one.

What do you think?

File: blockchain/chain/src/store/chain_store.rs [hunk 1]
```diff
@@ -523,38 +523,45 @@ where
...
+                    .map_err(|e| Error::Other(format!("Failed to write blocks in export: {e}")))
+            } else {
+                while rx.recv_async().await.is_ok() {}
+                Ok(())
+            }
         });
 
         let global_pre_time = SystemTime::now();
-        info!("chain export started");
+        info!("chain export started, dry_run: {dry_run}");
```
[the remark this answers is not in the document]
Comment (hanabi1224) [reply blockchain/chain/src/store/chain_store.rs]: fixed

**Comment by hanabi1224** [comment]
> Have you considered adding a dummy writer to the export method? 

Fixed.

File: utils/forest_utils/src/io/writer_checksum.rs [hunk 2]
```diff
@@ -67,6 +67,42 @@ impl<D: Digest, W> AsyncWriterWithChecksum<D, W> {
     }
 }
 
+/// A void writer that does nothing but implements [AsyncWrite] and [Checksum]
+#[derive(Debug, Clone, Default)]
+pub struct VoidAsyncWriterWithChecksum<D> {
```
Comment (LesnyRumcajs) [inline utils/forest_utils/src/io/writer_checksum.rs:72]: We could make it not so dry. I mean, don't save anything but still generate a correct checksum. No strong opinion, though.

**Review by LesnyRumcajs — APPROVED** [review]

File: utils/forest_utils/src/io/writer_checksum.rs (hunk 2 above)
Comment (hanabi1224) [reply utils/forest_utils/src/io/writer_checksum.rs:72]: Maybe rename it to `VoidAsyncWriterWithVoidChecksum`? For this particular case we don't really need the checksum, and we could implement what u suggest with `VoidAsyncWriterWithChecksum` when we need it. Sounds good?

Comment (hanabi1224) [reply utils/forest_utils/src/io/writer_checksum.rs]: Fixed.

**Review by tyshko5 — APPROVED** [review]

**Pull request merged**